File: backend/src/ssm/ssh/safety.py
"""Enforce host-level safety guards before any SSH write.

Two layers:

1. ``host.disabled`` flag from the database — checked before we even open a
   connection (see :func:`ensure_host_not_disabled`). Disabled hosts are a
   pure-Python signal; we never phone home to find this out.

2. Readonly sentinels on the remote host — ``~/.ssh/system_readonly`` (global
   for the SSH user) and the login user's ``~/.ssh/user_readonly`` (per-login).
   Non-empty contents of either file freeze writes and the contents become the
   reason surfaced to the UI.
"""

from __future__ import annotations

import shlex

from ssm.core.errors import HostDisabled, SshReadOnly
from ssm.ssh.protocol import SshClient, SshTarget

_READONLY_SCRIPT = r"""
set -u
system="$HOME/.ssh/system_readonly"
if [ -s "$system" ]; then
    printf 'system_readonly: %s\n' "$(cat "$system")"
    exit 0
fi
home_dir="$(getent passwd "$1" | cut -d: -f6)"
if [ -n "$home_dir" ]; then
    user_file="$home_dir/.ssh/user_readonly"
    if [ -s "$user_file" ]; then
        printf 'user_readonly: %s\n' "$(cat "$user_file")"
        exit 0
    fi
fi
"""


def ensure_host_not_disabled(*, disabled: bool, host_name: str) -> None:
    """Raise :class:`HostDisabled` when ``disabled`` is ``True``."""
    if disabled:
        raise HostDisabled(f"host {host_name!r} is disabled")


async def check_readonly(client: SshClient, target: SshTarget, login: str) -> str | None:
    """Return the readonly reason for ``login`` on ``target``, or ``None``."""
    command = f"sh -c {shlex.quote(_READONLY_SCRIPT)} sh {shlex.quote(login)}"
    result = await client.exec(target, command)
    reason = result.stdout.strip()
    return reason or None


async def ensure_writable(client: SshClient, target: SshTarget, login: str) -> None:
    """Raise :class:`SshReadOnly` if either readonly sentinel is set."""
    reason = await check_readonly(client, target, login)
    if reason is not None:
        raise SshReadOnly(f"host {target.name!r} is read-only for {login!r}: {reason}")

File: phoenix/lib/ssm/ssh/safety.ex
defmodule Ssm.Ssh.Safety do
  @moduledoc """
  Host-level safety guards before any SSH write — port of ssh/safety.py.

  1. `host.disabled` from the database: checked before opening a connection.
  2. Readonly sentinels on the remote host: `~/.ssh/system_readonly` (global
     for the SSH user) and the login's `~/.ssh/user_readonly`. Non-empty
     contents freeze writes; the contents become the reason shown in the UI.
  """

  alias Ssm.Ssh.{Result, Shell, Target}

  @readonly_script ~S"""
  set -u
  system="$HOME/.ssh/system_readonly"
  if [ -s "$system" ]; then
      printf 'system_readonly: %s\n' "$(cat "$system")"
      exit 0
  fi
  home_dir="$(getent passwd "$1" | cut -d: -f6)"
  if [ -n "$home_dir" ]; then
      user_file="$home_dir/.ssh/user_readonly"
      if [ -s "$user_file" ]; then
          printf 'user_readonly: %s\n' "$(cat "$user_file")"
          exit 0
      fi
  fi
  """

  @doc "Error when the host is disabled; the DB flag is the only source."
  @spec ensure_host_not_disabled(%{disabled: boolean(), name: String.t()}) ::
          :ok | {:error, {:host_disabled, String.t()}}
  def ensure_host_not_disabled(%{disabled: true, name: name}),
    do: {:error, {:host_disabled, "host #{inspect(name)} is disabled"}}

  def ensure_host_not_disabled(_host), do: :ok

  @doc "The readonly reason for `login` on `target`, or nil."
  @spec check_readonly(module(), Target.t(), String.t()) ::
          {:ok, String.t() | nil} | {:error, term()}
  def check_readonly(client, target, login) do
    command = "sh -c #{Shell.quote(@readonly_script)} sh #{Shell.quote(login)}"

    with {:ok, %Result{} = result} <- client.exec(target, command, []) do
      case String.trim(result.stdout) do
        "" -> {:ok, nil}
        reason -> {:ok, reason}
      end
    end
  end

  @doc "Error with the readonly reason if either sentinel is set."
  @spec ensure_writable(module(), Target.t(), String.t()) :: :ok | {:error, term()}
  def ensure_writable(client, target, login) do
    case check_readonly(client, target, login) do
      {:ok, nil} ->
        :ok

      {:ok, reason} ->
        {:error,
         {:ssh_readonly,
          "host #{inspect(target.name)} is read-only for #{inspect(login)}: #{reason}"}}

      {:error, _} = error ->
        error
    end
  end
end
